pkg/image/api/validation/validation.go

```diff
@@ -59,9 +60,74 @@ func validateImage(image *api.Image, fldPath *field.Path) field.ErrorList {
 		}
 	}
 
+	for i, sig := range image.Signatures {
+		result = append(result, validateImageSignature(&sig, fldPath.Child("signatures").Index(i))...)
+	}
+
 	return result
 }
 
+func validateImageSignature(signature *api.ImageSignature, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(signature.Type) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("type"), ""))
+	}
+	if len(signature.Content) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("content"), ""))
+	}
+
+	var (
+		trustedCondition, forImageCondition *api.SignatureCondition
+		forImageConditionIndex              = -1
+	)
+	for i := range signature.Conditions {
+		cond := &signature.Conditions[i]
+		if cond.Type == api.SignatureTrusted && (trustedCondition == nil || !cond.LastProbeTime.Before(trustedCondition.LastProbeTime)) {
+			trustedCondition = cond
+		} else if cond.Type == api.SignatureForImage && forImageCondition == nil || !cond.LastProbeTime.Before(forImageCondition.LastProbeTime) {
+			forImageCondition = cond
+			forImageConditionIndex = i
+		}
+	}
+
+	if trustedCondition != nil && forImageCondition == nil {
+		msg := fmt.Sprintf("missing %q condition type", api.SignatureForImage)
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), signature.Conditions, msg))
+	} else if forImageCondition != nil && trustedCondition == nil {
+		msg := fmt.Sprintf("missing %q condition type", api.SignatureTrusted)
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), signature.Conditions, msg))
+	} else if trustedCondition != nil && forImageCondition != nil {
+		if ((trustedCondition.Status == kapi.ConditionUnknown) && (forImageCondition.Status != kapi.ConditionUnknown)) || ((trustedCondition.Status == kapi.ConditionTrue) && (forImageCondition.Status != kapi.ConditionTrue)) {
```

Comment:
Upstream has very relaxed conditions validation. In fact, they don't validate them at all. I removed this check.
On 06/06/2016 05:30 PM, David Eads wrote:
>
> In pkg/image/api/validation/validation.go <https://github.com/openshift/origin/pull/8371#discussion_r65911813>:
>
> > +		if cond.Type == api.SignatureTrusted && (trustedCondition == nil || !cond.LastProbeTime.Before(trustedCondition.LastProbeTime)) {
> > +			trustedCondition = cond
> > +		} else if cond.Type == api.SignatureForImage && forImageCondition == nil || !cond.LastProbeTime.Before(forImageCondition.LastProbeTime) {
> > +			forImageCondition = cond
> > +			forImageConditionIndex = i
> > +		}
> > +	}
> > +
> > +	if trustedCondition != nil && forImageCondition == nil {
> > +		msg := fmt.Sprintf("missing %q condition type", api.SignatureForImage)
> > +		allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), signature.Conditions, msg))
> > +	} else if forImageCondition != nil && trustedCondition == nil {
> > +		msg := fmt.Sprintf("missing %q condition type", api.SignatureTrusted)
> > +		allErrs = append(allErrs, field.Invalid(fldPath.Child("conditions"), signature.Conditions, msg))
> > +	} else if trustedCondition != nil && forImageCondition != nil {
> > +		if ((trustedCondition.Status == kapi.ConditionUnknown) && (forImageCondition.Status != kapi.ConditionUnknown)) || ((trustedCondition.Status == kapi.ConditionTrue) && (forImageCondition.Status != kapi.ConditionTrue)) {
>
> I think you need to allow this condition combination. Otherwise, you cannot produce conditions for "you're under 
> attack". The "Trust" in the sense is just that we recognize and respect the signer, not that he gave us something good.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/openshift/origin/pull/8371/files/4371c2172d2421e632faa67cf851acce66753c04#r65911813>, or mute the 
> thread <https://github.com/notifications/unsubscribe/ABaQk4oC_fWPtuHllwFzDddUPvL1Lbkoks5qJD0ogaJpZM4IBAFQ>.
>